### PR TITLE
Compute onboarding state – WEB-287

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description
A pragmatic approach to compute the state of the cell to tell if the Dedicated Onboarding flow should be shown.

This adds `getOnboardingState` to server which result resolves to 
```ts
    interface OnboardingState {
        /**
         * Whether this Gitpod instance is already configured with SSO.
         */
        readonly isCompleted: boolean;

        /**
         * Whether this Gitpod instance has at least one org.
         */
        readonly hasAnyOrg: boolean;
    }
```

## Related Issue(s)

Fixes WEB-287

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - at-onboarding-state</li>
	<li><b>🔗 URL</b> - <a href="https://at-onboarding-state.preview.gitpod-dev.com/workspaces" target="_blank">at-onboarding-state.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [x] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

/hold

### CODE REVIEW

File: components/gitpod-db/src/typeorm/team-db-impl.ts [hunk 1]
```diff
@@ -66,6 +66,8 @@ export class TeamDBImpl implements TeamDB {
         const queryBuilder = teamRepo
             .createQueryBuilder("team")
             .where("LOWER(team.name) LIKE LOWER(:searchTerm)", { searchTerm: `%${searchTerm}%` })
+            .andWhere("deleted = 0")
```
Comment (AlexTugarev) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:69]: drive-by fix: `findTeams` did not respect deletion flag at all.

File: components/gitpod-protocol/src/gitpod-service.ts [hunk 2]
```diff
@@ -473,6 +476,22 @@ export namespace GitpodServer {
...
+        /**
+         * Whether any org is .
+         */
+        readonly hasAnyOrg: boolean;
+
+        /**
+         * Whether this
+         */
+        readonly hasAnyOrgWithInactiveSSO: boolean;
+    }
```
Comment (easyCZ) [inline components/gitpod-protocol/src/gitpod-service.ts:494]: While reviewing this, I considered we could return the actual counts instead of the booleans, but that also presents challenges for SaaS, so I think the above opaque is a good balance.

File: components/server/src/workspace/gitpod-server-impl.ts [hunk 3]
```diff
@@ -3980,6 +3980,54 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
...
+        // Optimized check for completed setup
+        const hasAnyOrgWithActiveSSO = await this.teamDB.hasAnyOrgWithActiveSSO();
+        if (hasAnyOrgWithActiveSSO) {
+            // Let's cache the state at least per connection.
+            this._getOnboardingState = {
+                isCompleted: true,
+                hasAnyOrg: true,
+                hasAnyOrgWithInactiveSSO: false,
+            };
+            return this._getOnboardingState;
```
Comment (easyCZ) [inline components/server/src/workspace/gitpod-server-impl.ts:4011]: Why? This would only really be encountered during the onboarding flow, and never after, no?

File: components/gitpod-protocol/src/gitpod-service.ts [hunk 4]
```diff
@@ -473,6 +476,22 @@ export namespace GitpodServer {
...
+        readonly isCompleted: boolean;
+
+        /**
+         * Whether any org is .
+         */
+        readonly hasAnyOrg: boolean;
+
+        /**
+         * Whether this
+         */
```
Comment (easyCZ) [inline components/gitpod-protocol/src/gitpod-service.ts:492]: Drop comment, or complete it

File: components/gitpod-db/src/typeorm/team-db-impl.ts [hunk 5]
```diff
@@ -377,10 +379,14 @@ export class TeamDBImpl implements TeamDB {
         }
     }
 
-    public async someOrgWithSSOExists(): Promise<boolean> {
-        const repo = await this.getOrgSettingsRepo();
+    public async hasAnyOrgWithActiveSSO(): Promise<boolean> {
+        const repo = await this.getTeamRepo();
         const result = await repo.query(
-            `select org.id from d_b_team as org inner join d_b_oidc_client_config as oidc on org.id = oidc.organizationId limit 1;`,
+            `select org.id from d_b_team as org inner join d_b_oidc_client_config as oidc on org.id = oidc.organizationId
```
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:385]: A bit unfortunate that server now talks to the OIDC table, which is owned by public api. We'll follow-up and clean-up this once we know better exactly how we can discover that state

File: components/gitpod-protocol/src/gitpod-service.ts [hunk 6]
```diff
@@ -473,6 +476,22 @@ export namespace GitpodServer {
...
+        readonly hasAnyOrgWithInactiveSSO: boolean;
```
Comment (selfcontained) [inline components/gitpod-protocol/src/gitpod-service.ts:493]: Should this be `hasAnyOrgWithActiveSSO` instead?

File: components/server/src/workspace/gitpod-server-impl.ts (hunk 3 above)
Comment (selfcontained) [reply components/server/src/workspace/gitpod-server-impl.ts:4011]: This call will happen every time we load the dashboard for a logged in user, which is how it can determine if a setup flow needs to be presented.

File: components/gitpod-db/src/typeorm/team-db-impl.ts (hunk 5 above)
Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:385]: Yeah! The whole point to opt for server as home of this, is that it was initially called from `getLoggedInUser`. Now that this isn't the case, we can actually move it to papi. 🤔

File: components/gitpod-protocol/src/gitpod-service.ts (hunk 6 above)
Comment (AlexTugarev) [reply components/gitpod-protocol/src/gitpod-service.ts:493]: > Should this be hasAnyOrgWithActiveSSO instead?

Wouldn't that be equal to `isCompleted`? 

I recall, you'd be interested in the state of having any Org without verified SSO.

File: components/server/src/workspace/gitpod-server-impl.ts (hunk 3 above)
Comment (AlexTugarev) [reply components/server/src/workspace/gitpod-server-impl.ts:4011]: Yes. The motivation is to reduce load/delays by this calls, as the frontend would make them on _every_ initial load. There we can optimize as well, and memorize whether onboarding is completed.

File: components/gitpod-protocol/src/gitpod-service.ts (hunk 6 above)
Comment (selfcontained) [reply components/gitpod-protocol/src/gitpod-service.ts:493]: Good point, I suppose I'm not sure how we'd use this then, we can drop for now and add later if we have a reason.

Comment (AlexTugarev) [reply components/gitpod-protocol/src/gitpod-service.ts:493]: Thought about it, it seems to be redundant, or at least less helpful information. 

If you know that the onboarding is incomplete, and there is at least one org, you'd try to fetch the org, and eventually the SSO configs by org ID.

Comment (AlexTugarev) [reply components/gitpod-protocol/src/gitpod-service.ts]: Dropped it.

File: components/gitpod-db/src/typeorm/team-db-impl.ts (hunk 5 above)
Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:385]: I'll look into rewriting in papi this one.

Comment (easyCZ) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:385]: @AlexTugarev Let's keep it like this. I think eventually this should be solvable with general APIs and proper Authorization which can grant the admin user access to all orgs to query.
